src/libsyntax/ext/expand.rs

```diff
@@ -188,12 +189,30 @@ impl Invocation {
 
 pub struct MacroExpander<'a, 'b:'a> {
     pub cx: &'a mut ExtCtxt<'b>,
+    partial_expansions: HashMap<Mark, PartialExpansion>,
+    full_expansions: HashMap<Mark, FullExpansion>,
     monotonic: bool, // c.f. `cx.monotonic_expander()`
 }
 
+struct PartialExpansion {
```

Comment:
Will do. Partial expansion are macro expansions that have unexpanded macro invocations in them. That is, the macro itself has resolved and been expanded, but it created more macro invocations that have yet to be expanded. This is in contrast to "fully expanded" AST, which has no macro invocations left.